lenskit-predict/src/test/java/org/grouplens/lenskit/predict/KnownRatingRatingPredictorTest.java

```diff
@@ -0,0 +1,129 @@
+    //Setup method
+    public void createPredictor() throws RecommenderBuildException {
+        rs.add(Ratings.make(14, 1, 5));
+        rs.add(Ratings.make(14, 2, 4));
+        rs.add(Ratings.make(14, 3, 3));
+        rs.add(Ratings.make(14, 4, 2));
+        rs.add(Ratings.make(14, 5, 0.2));
+        rs.add(Ratings.make(15, 5, 1));
+        rs.add(Ratings.make(15, 6, 2));
+        rs.add(Ratings.make(15, 7, 3));
+        rs.add(Ratings.make(15, 8, 4));
+        rs.add(Ratings.make(15, 9, 5));
+
+        dao = new EventCollectionDAO(rs);
+        userDAO = new PrefetchingUserEventDAO(dao);
+    }
+
+    @Test
+    /*
+    * Test method that tests predicting for a user not in the data set
+    * */
+    public void testPredictForMissingUser() {
+        KnownRatingRatingPredictor KnownPredict = new KnownRatingRatingPredictor(userDAO);
+        MutableSparseVector predictItems = MutableSparseVector.create(1, 2);
+        KnownPredict.predict(5, predictItems);
+        assertThat(predictItems.size(), equalTo(0));
+    }
+
+    @Test
+     /*
+    * Test method that tests predicting for a user in the data set,
+    * only with items they have rated.
+    * */
+    public void testPredictForRatingByGivenUser() {
+        KnownRatingRatingPredictor KnownPredict = new KnownRatingRatingPredictor(userDAO);
+        long[] keys= {1,3,5};
+        double[] values= {1.0,2.0,4.0};
+        MutableSparseVector predictItems = MutableSparseVector.wrap(keys, values);
+        KnownPredict.predict(14, predictItems);
+        assertThat(predictItems.get(1), equalTo(5.0));
+        assertThat(predictItems.get(3), equalTo(3.0));
+        assertThat(predictItems.get(5), equalTo(0.2));
+    }
+
+    @Test
+     /*
+    * Test method that tests predicting for a user in the data set,
+    * returning the rating of item for given user only.
+    * */
+    public void  testPredictForOnlyRatedItems() {
+        KnownRatingRatingPredictor KnownPredict = new KnownRatingRatingPredictor(userDAO);
+        MutableSparseVector predictItems = MutableSparseVector.create(5, 7, 1);
+        KnownPredict.predict(15, predictItems);
+        assertThat(predictItems.get(5), equalTo(1.0));
+        assertThat(predictItems.get(7), equalTo(3.0));
+        assertFalse(predictItems.containsKey(1));
+
+    }
+    @Test
+     /*
+    * Test method that tests unrated items for a user in the data set,
+    * it shouldn't return any value.
+    * */
+    public void  testPredictForUnratedItems() {
+
+        Rating r = new RatingBuilder()
+                .setUserId(420)
+                .setItemId(840)
+                .setRating(3.5)
+                .clearRating()
+                .build();
+        rs.add(r);
+
+        KnownRatingRatingPredictor KnownPredict = new KnownRatingRatingPredictor(userDAO);
+        MutableSparseVector predictItems = MutableSparseVector.create(840);
+        KnownPredict.predict(420, predictItems);
+        assertThat(predictItems.size(), equalTo(0));
```

Comment:
Good start. But this test doesn't quite test the behavior we're looking for; we need an unrate event *that undoes a previous rating*. For this, we need the ratings to actually have timestamps (rather than -1), and we need 2 events (and I would suggest 3, so that we can verify that it doesn't stomp on other ratings.  So something like:

```
RatingBuilder rb = new RatingBuilder().setUserId(420);
rs.add(rb.setItemId(840).setRating(3.5).setTimestamp(10));
rs.add(rb.setItemId(390).setRating(4.5).setTimestamp(20));
rs.add(rb.setItemId(840).clearRating().setTimestamp(30));
```

Also, we need to re-construct the DAO as well, because it takes a copy of the rating list.

After these changes, I think it will be ready to merge! Good work.